File: frontend/campus-dashboard/src/lib/firestore-chat.js
/**
 * Firestore real-time chat client for the AI assistant widget.
 *
 * Messages are stored in ``chat_sessions/{uid}/messages/``.
 * The frontend subscribes with onSnapshot() for real-time updates,
 * giving a streaming-like experience without WebSockets.
 *
 * Used by the Campus Assistant chat widget as an alternative transport
 * when Firestore is available.
 */
import { initializeApp, getApps } from 'firebase/app';
import {
  getFirestore,
  collection,
  query,
  orderBy,
  limit as fbLimit,
  onSnapshot,
  doc,
  setDoc,
  increment,
  deleteDoc,
  getDocs,
} from 'firebase/firestore';

/** Shared Firebase web config. */
function getConfig() {
  const apiKey = import.meta.env.VITE_FIREBASE_API_KEY;
  const projectId = import.meta.env.VITE_FIREBASE_PROJECT_ID;
  if (!apiKey || !projectId) return null;
  return {
    apiKey,
    authDomain: import.meta.env.VITE_FIREBASE_AUTH_DOMAIN || `${projectId}.firebaseapp.com`,
    projectId,
    messagingSenderId: import.meta.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
    appId: import.meta.env.VITE_FIREBASE_APP_ID,
  };
}

let _app = null;
let _db = null;

function getDb() {
  if (_db) return _db;
  const config = getConfig();
  if (!config) return null;

  _app = getApps().length
    ? getApps()[0]
    : initializeApp(config);

  _db = getFirestore(_app);
  return _db;
}

function _messagesCol(userUid) {
  const db = getDb();
  if (!db) return null;
  return collection(db, 'chat_sessions', userUid, 'messages');
}

function _sessionRef(userUid) {
  const db = getDb();
  if (!db) return null;
  return doc(db, 'chat_sessions', userUid);
}

/**
 * Subscribe to real-time chat messages for a user.
 *
 * @param {string} userUid
 * @param {function} callback — called with [{ role, content, created_at }]
 * @param {number} [maxMessages=50]
 * @returns {function} unsubscribe
 */
export function subscribeToChat(userUid, callback, maxMessages = 50) {
  const col = _messagesCol(userUid);
  if (!col) return () => {};

  const q = query(col, orderBy('created_at', 'asc'), fbLimit(maxMessages));

  return onSnapshot(q, (snapshot) => {
    const messages = snapshot.docs.map((d) => ({
      id: d.id,
      ...d.data(),
      created_at: d.data().created_at?.toDate?.()?.toISOString() || null,
    }));
    callback(messages);
  });
}

/**
 * Send a user message to Firestore.
 */
export async function sendUserMessage(userUid, content) {
  const col = _messagesCol(userUid);
  if (!col) return null;

  try {
    const msgRef = doc(col);
    await setDoc(msgRef, {
      role: 'user',
      content,
      created_at: new Date(),
    });

    // Update session metadata.
    const sessionRef = _sessionRef(userUid);
    if (sessionRef) {
      await setDoc(sessionRef, {
        updated_at: new Date(),
        message_count: increment(1),
      }, { merge: true });
    }

    return msgRef.id;
  } catch (err) {
    console.warn('Firestore sendUserMessage failed:', err);
    return null;
  }
}

/**
 * Send an assistant reply to Firestore.
 */
export async function sendAssistantMessage(userUid, content) {
  const col = _messagesCol(userUid);
  if (!col) return null;

  try {
    const msgRef = doc(col);
    await setDoc(msgRef, {
      role: 'assistant',
      content,
      created_at: new Date(),
    });

    const sessionRef = _sessionRef(userUid);
    if (sessionRef) {
      await setDoc(sessionRef, {
        updated_at: new Date(),
        message_count: increment(1),
      }, { merge: true });
    }

    return msgRef.id;
  } catch (err) {
    console.warn('Firestore sendAssistantMessage failed:', err);
    return null;
  }
}

/**
 * Clear all chat messages for a user.
 */
export async function clearChatSession(userUid) {
  const col = _messagesCol(userUid);
  if (!col) return;

  try {
    const snapshot = await getDocs(col);
    const deletes = snapshot.docs.map((d) => deleteDoc(d.ref));
    await Promise.all(deletes);

    const sessionRef = _sessionRef(userUid);
    if (sessionRef) {
      await deleteDoc(sessionRef).catch(() => {});
    }
  } catch (err) {
    console.warn('Firestore clearChatSession failed:', err);
  }
}

File: backend/campus_project/firestore_chat.py
"""Firestore-backed real-time chat store for the AI assistant.

Messages are stored per-user in a sub-collection:

    chat_sessions/{uid}/messages/{auto-id}
      role: "user" | "assistant"
      content: str
      created_at: Timestamp

The top-level document tracks session metadata:

    chat_sessions/{uid}
      created_at: Timestamp
      updated_at: Timestamp
      message_count: int

The frontend can ``onSnapshot()`` the ``messages`` sub-collection to
get real-time streaming-like updates without WebSockets.
"""

import logging
from datetime import datetime, timezone

logger = logging.getLogger(__name__)


def _get_client():
    from campus_project.firestore_client import get_firestore
    return get_firestore()


def _session_col(user_uid):
    db = _get_client()
    if db is None:
        return None
    return db.collection('chat_sessions').document(str(user_uid))


def save_message(user_uid, role, content):
    """Append a message to the user's chat session.

    ``role`` is ``'user'`` or ``'assistant'``.
    """
    db = _get_client()
    if db is None:
        return None

    now = datetime.now(tz=timezone.utc)
    session_ref = _session_col(user_uid)
    if session_ref is None:
        return None

    try:
        msg_ref = session_ref.collection('messages').document()
        msg_ref.set({
            'role': role,
            'content': content,
            'created_at': now,
        })

        # Update session metadata.
        session_ref.set({
            'created_at': now,
            'updated_at': now,
            'message_count': _firestore.Increment(1),
        }, merge=True)

        return msg_ref.id
    except Exception:
        logger.debug('Firestore save_message failed.', exc_info=True)
        return None


def get_recent_messages(user_uid, limit=20):
    """Return the most recent messages for context window."""
    db = _get_client()
    if db is None:
        return []

    try:
        session_ref = _session_col(user_uid)
        if session_ref is None:
            return []

        query = (
            session_ref.collection('messages')
            .order_by('created_at', direction='DESCENDING')
            .limit(limit)
        )
        messages = []
        for snap in query.stream():
            data = snap.to_dict()
            messages.append({
                'role': data.get('role', 'user'),
                'content': data.get('content', ''),
            })
        messages.reverse()  # oldest first for context
        return messages
    except Exception:
        logger.debug('Firestore get_recent_messages failed.', exc_info=True)
        return []


def clear_session(user_uid):
    """Delete all messages in a user's chat session."""
    db = _get_client()
    if db is None:
        return

    try:
        session_ref = _session_col(user_uid)
        if session_ref is None:
            return

        docs = session_ref.collection('messages').stream()
        batch = db.batch()
        count = 0
        for doc in docs:
            batch.delete(doc.reference)
            count += 1
            if count % 500 == 0:
                batch.commit()
                batch = db.batch()
        if count % 500 != 0:
            batch.commit()

        session_ref.delete()
    except Exception:
        logger.debug('Firestore clear_session failed.', exc_info=True)


def subscribe_to_messages(user_uid, callback):
    """Register a real-time listener for new messages.

    ``callback`` receives a list of message dicts each time the
    collection changes.  Returns an unsubscribe handle.

    Only works client-side (JS SDK).  For the Python backend this is
    a no-op — the backend writes, the frontend subscribes.
    """
    # Firestore real-time listeners are a client-side JS feature.
    # This is a placeholder documenting the intended API contract.
    return lambda: None
